File: app/src/main/java/com/elementary/tasks/core/utils/SharedPrefs.java
/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.elementary.tasks.core.utils;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.Gson;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.Map;

abstract class SharedPrefs extends PrefsConstants {

    private SharedPreferences prefs;

    private SharedPrefs() {
    }

    SharedPrefs(Context context){
        prefs = context.getSharedPreferences("ui_settings", Context.MODE_PRIVATE);
    }

    void putString(String stringToSave, String value){
        prefs.edit().putString(stringToSave, value).apply();
    }

    void putInt(String stringToSave, int value){
        prefs.edit().putInt(stringToSave, value).apply();
    }

    int getInt(String stringToLoad){
        int x;
        try {
            x = prefs.getInt(stringToLoad, 0);
        } catch (ClassCastException e) {
            try {
                x = Integer.parseInt(prefs.getString(stringToLoad, "0"));
            } catch (ClassCastException e1) {
                x = 0;
            }
        }
        return x;
    }

    void putLong(String stringToSave, long value){
        prefs.edit().putLong(stringToSave, value).apply();
    }

    long getLong(String stringToLoad){
        long x;
        try {
            x = prefs.getLong(stringToLoad, 1000);
        } catch (ClassCastException e) {
            x = Long.parseLong(prefs.getString(stringToLoad, "1000"));
        }
        return x;
    }

    void putObject(String key, Object obj) {
        Gson gson = new Gson();
        putString(key, gson.toJson(obj));
    }

    Object getObject(String key, Class<?> classOfT) {
        String json = getString(key);
        Object value = new Gson().fromJson(json, classOfT);
        if (value == null)
            throw new NullPointerException();
        return value;
    }

    String getString(String stringToLoad){
        String res;
        try {
            res = prefs.getString(stringToLoad, "");
        } catch (NullPointerException e) {
            e.printStackTrace();
            res = "";
        }
        return res;
    }

    boolean hasKey(String checkString){
        return prefs.contains(checkString);
    }

    void putBoolean(String stringToSave, boolean value){
        prefs.edit().putBoolean(stringToSave, value).apply();
    }

    boolean getBoolean(String stringToLoad){
        boolean res;
        try {
            res = prefs.getBoolean(stringToLoad, false);
        } catch (ClassCastException e){
            res = Boolean.parseBoolean(prefs.getString(stringToLoad, "false"));
        }
        return res;
    }

    void saveVersionBoolean(String stringToSave){
        prefs.edit().putBoolean(stringToSave, true).apply();
    }

    boolean getVersion(String stringToLoad){
        boolean res;
        try {
            res = prefs.getBoolean(stringToLoad, false);
        } catch (ClassCastException e){
            res = Boolean.parseBoolean(prefs.getString(stringToLoad, "false"));
        }
        return res;
    }

    void savePrefsBackup(String key){
        File dir = MemoryUtil.getPrefsDir();
        if (dir != null) {
            File prefsFile = new File(dir + "/prefs.xml");
            if (prefsFile.exists()) prefsFile.delete();
            ObjectOutputStream output = null;
            try {
                output = new ObjectOutputStream(new FileOutputStream(prefsFile));
                Map<String, ?> list = prefs.getAll();
                if (list.containsKey(key)) list.remove(key);
                output.writeObject(list);
            } catch (IOException e) {
                e.printStackTrace();
            }finally {
                try {
                    if (output != null) {
                        output.flush();
                        output.close();
                    }
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
            }
        }
    }

    void loadPrefsFromFile(){
        File dir = MemoryUtil.getPrefsDir();
        if (dir == null) return;

        File prefsFile = new File(dir + "/prefs.xml");
        if (prefsFile.exists()) {
            ObjectInputStream input = null;
            try {
                input = new ObjectInputStream(new FileInputStream(prefsFile));
                SharedPreferences.Editor prefEdit = prefs.edit();
                prefEdit.clear();
                Map<String, ?> entries = (Map<String, ?>) input.readObject();
                for (Map.Entry<String, ?> entry : entries.entrySet()) {
                    Object v = entry.getValue();
                    String key = entry.getKey();
                    if (v instanceof Boolean)
                        prefEdit.putBoolean(key, (Boolean) v);
                    else if (v instanceof Float)
                        prefEdit.putFloat(key, (Float) v);
                    else if (v instanceof Integer)
                        prefEdit.putInt(key, (Integer) v);
                    else if (v instanceof Long)
                        prefEdit.putLong(key, (Long) v);
                    else if (v instanceof String)
                        prefEdit.putString(key, ((String) v));
                }
                prefEdit.apply();
            } catch (ClassNotFoundException | IOException e) {
                e.printStackTrace();
            } finally {
                try {
                    if (input != null) {
                        input.close();
                    }
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
            }
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/utils/SharedPrefs.kt
package com.elementary.tasks.core.utils

import android.content.Context
import android.content.SharedPreferences
import com.elementary.tasks.core.cloud.FileConfig
import com.google.gson.Gson
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

abstract class SharedPrefs(protected val context: Context) {
  private var prefs: SharedPreferences = context.getSharedPreferences(PrefsConstants.PREFS_NAME, Context.MODE_PRIVATE)

  fun getLongArray(stringToLoad: String): Array<Long> {
    return try {
      prefs.getStringSet(stringToLoad, setOf<String>())?.map {
        try {
          it.toLong()
        } catch (e: Exception) {
          0L
        }
      }?.toTypedArray() ?: arrayOf()
    } catch (e: Exception) {
      arrayOf()
    }
  }

  fun putLongArray(stringToSave: String, array: Array<Long>) {
    prefs.edit().putStringSet(stringToSave, array.map { it.toString() }.toSet()).apply()
  }

  fun putStringArray(stringToSave: String, array: Array<String>) {
    prefs.edit().putStringSet(stringToSave, array.toSet()).apply()
  }

  fun getStringArray(stringToLoad: String): Array<String> {
    return try {
      prefs.getStringSet(stringToLoad, setOf<String>())?.toTypedArray() ?: arrayOf()
    } catch (e: Exception) {
      arrayOf()
    }
  }

  fun putString(stringToSave: String, value: String) {
    prefs.edit().putString(stringToSave, value).apply()
  }

  fun putInt(stringToSave: String, value: Int) {
    prefs.edit().putInt(stringToSave, value).apply()
  }

  fun getInt(stringToLoad: String, def: Int = 0): Int {
    return try {
      prefs.getInt(stringToLoad, def)
    } catch (e: ClassCastException) {
      try {
        Integer.parseInt(prefs.getString(stringToLoad, "$def") ?: "$def")
      } catch (e1: ClassCastException) {
        def
      }
    }
  }

  fun putLong(stringToSave: String, value: Long) {
    prefs.edit().putLong(stringToSave, value).apply()
  }

  fun getLong(stringToLoad: String): Long {
    return try {
      prefs.getLong(stringToLoad, 1000)
    } catch (e: ClassCastException) {
      java.lang.Long.parseLong(prefs.getString(stringToLoad, "1000") ?: "1000")
    }
  }

  fun putObject(key: String, obj: Any) {
    putString(key, Gson().toJson(obj))
  }

  fun getString(stringToLoad: String): String {
    return prefs.getString(stringToLoad, "") ?: ""
  }

  fun hasKey(checkString: String): Boolean {
    return prefs.contains(checkString)
  }

  fun removeKey(checkString: String) {
    prefs.edit().remove(checkString).apply()
  }

  fun putBoolean(stringToSave: String, value: Boolean) {
    prefs.edit().putBoolean(stringToSave, value).apply()
  }

  fun getBoolean(stringToLoad: String, def: Boolean = false): Boolean {
    return try {
      prefs.getBoolean(stringToLoad, def)
    } catch (e: ClassCastException) {
      java.lang.Boolean.parseBoolean(prefs.getString(stringToLoad, "false"))
    }
  }

  fun saveVersionBoolean(stringToSave: String) {
    prefs.edit().putBoolean(stringToSave, true).apply()
  }

  fun getVersion(stringToLoad: String): Boolean {
    return try {
      prefs.getBoolean(stringToLoad, false)
    } catch (e: ClassCastException) {
      java.lang.Boolean.parseBoolean(prefs.getString(stringToLoad, "false"))
    }
  }

  fun all(): Map<String, *> {
    return prefs.all
  }

  fun sharedPrefs(): SharedPreferences {
    return prefs
  }

  fun savePrefsBackup() {
    val dir = MemoryUtil.prefsDir
    if (dir != null) {
      val prefsFile = File("$dir/" + FileConfig.FILE_NAME_SETTINGS)
      if (prefsFile.exists()) {
        prefsFile.delete()
      }
      var output: ObjectOutputStream? = null
      try {
        output = ObjectOutputStream(FileOutputStream(prefsFile))
        val list = prefs.all
        if (list.containsKey(PrefsConstants.DRIVE_USER)) {
          list.remove(PrefsConstants.DRIVE_USER)
        }
        if (list.containsKey(PrefsConstants.TASKS_USER)) {
          list.remove(PrefsConstants.TASKS_USER)
        }
        output.writeObject(list)
      } catch (e: IOException) {
        e.printStackTrace()
      } finally {
        try {
          if (output != null) {
            output.flush()
            output.close()
          }
        } catch (ex: IOException) {
          ex.printStackTrace()
        }
      }
    }
  }

  fun loadPrefsFromFile() {
    val dir = MemoryUtil.prefsDir ?: return
    val prefsFile = File("$dir/" + FileConfig.FILE_NAME_SETTINGS)
    if (prefsFile.exists()) {
      var input: ObjectInputStream? = null
      try {
        input = ObjectInputStream(FileInputStream(prefsFile))
        val prefEdit = prefs.edit()
        val entries = input.readObject() as Map<String, *>
        for ((key, v) in entries) {
          when (v) {
            is Boolean -> prefEdit.putBoolean(key, v)
            is Float -> prefEdit.putFloat(key, v)
            is Int -> prefEdit.putInt(key, v)
            is Long -> prefEdit.putLong(key, v)
            is String -> prefEdit.putString(key, v)
          }
        }
        prefEdit.apply()
      } catch (e: ClassNotFoundException) {
        e.printStackTrace()
      } catch (e: IOException) {
        e.printStackTrace()
      } finally {
        try {
          input?.close()
        } catch (ex: IOException) {
          ex.printStackTrace()
        }

      }
    }
  }
}